golang/cosmos/x/swingset/keeper/keeper.go

```diff
@@ -278,91 +334,90 @@ func (k Keeper) GetStorage(ctx sdk.Context, path string) string {
 	store := ctx.KVStore(k.storeKey)
-	dataStore := prefix.NewStore(store, types.DataPrefix)
-	keysStore := prefix.NewStore(store, types.KeysPrefix)
+	dataStore := prefix.NewStore(store, types.DataKeyPrefix)
+	dataKey := pathToDataKey(path)
 
-	// Update the value.
-	pathKey := stringToKey(path)
-	if value == "" {
-		dataStore.Delete(pathKey)
-	} else {
-		dataStore.Set(pathKey, []byte(value))
-	}
+	// Check if we have data.
+	return dataStore.Has(dataKey)
+}
+
+// HasKeys tells if a given path has child keys.
+func (k Keeper) HasKeys(ctx sdk.Context, path string) bool {
+	// Check if we have children.
+	iterator := k.getKeyIterator(ctx, path)
+	defer iterator.Close()
+	return iterator.Valid()
+}
 
+// SetStorage sets the entire generic storage for a path
+func (k Keeper) SetStorage(ctx sdk.Context, path, value string) {
 	ctx.EventManager().EmitEvent(
 		types.NewStorageEvent(path, value),
 	)
 
-	if value == "" && len(k.GetKeys(ctx, path).Keys) > 0 {
-		// If we're deleting and we had children, we need to keep our ancestor keys
-		// around.
-		return
+	store := ctx.KVStore(k.storeKey)
+	dataStore := prefix.NewStore(store, types.DataKeyPrefix)
+	pathStore := prefix.NewStore(store, types.PathKeyPrefix)
+
+	// Update the value.
+	dataKey := pathToDataKey(path)
+
+	if value == "" {
+		dataStore.Delete(dataKey)
+	} else {
+		dataStore.Set(dataKey, []byte(value))
 	}
 
-	// Update the parent keys.
+	// Update our and other parent keys.
 	pathComponents := strings.Split(path, ".")
-	for i := len(pathComponents) - 1; i >= 0; i-- {
+	for i := len(pathComponents); i >= 0; i-- {
 		ancestor := strings.Join(pathComponents[0:i], ".")
-		lastKeyStr := pathComponents[i]
-
-		// Get a map corresponding to the parent's keys.
-		keyNode := k.GetKeys(ctx, ancestor)
-		keyList := keyNode.Keys
-		keyMap := make(map[string]bool, len(keyList)+1)
-		for _, keyStr := range keyList {
-			keyMap[keyStr] = true
-		}
 
-		// Decide if we need to add or remove the key.
+		// Decide if we need to add or remove the ancestor.
 		if value == "" {
-			if _, ok := keyMap[lastKeyStr]; !ok {
-				// If the key is already gone, we don't need to remove from the key
-				// lists.
+			if k.HasStorage(ctx, ancestor) || k.HasKeys(ctx, ancestor) {
+				// If the key is needed, skip out.
 				return
 			}
 			// Delete the key.
-			delete(keyMap, lastKeyStr)
-		} else {
-			if _, ok := keyMap[lastKeyStr]; ok {
-				// If the key already exists, we don't need to add to the key lists.
-				return
-			}
-			// Add the key.
-			keyMap[lastKeyStr] = true
-		}
-
-		keyList = make([]string, 0, len(keyMap))
-		for keyStr := range keyMap {
-			keyList = append(keyList, keyStr)
-		}
-
-		// Update the list of keys
-		ancestorKey := stringToKey(ancestor)
-		if len(keyList) == 0 {
-			// No keys left, delete the parent.
-			keysStore.Delete(ancestorKey)
+			pathStore.Delete(pathToKey(ancestor))
+		} else if i < len(pathComponents) && k.HasStorage(ctx, ancestor) {
```

Comment:
This works, but looks like it unnecessarily re-creates storage-less path entries. Assuming the invariant:

```
path entries exist if and only if self or some descendant has non-empty storage
```

then any path entry can assume that all its ancestors also have path entries. So if there is some routine `k.HasPath(ctx, path)` to see if the path entry exists, you can use `HasPath` instead of `HasStorage` here.

Reply:
Great, thanks!